### PR TITLE
Feature: Copilot Chat

### PR DESCRIPTION
**Pull request by adrianwyatt**
- Adds frontend webapp (React)
- Add backend SKWebApi (SK REST API Service) for CopilotChat sample
- Adds CosmosDB connector memory store

### CODE REVIEW

File: samples/apps/github-qna-webapp-react/.env
Comment (dluc) [inline samples/apps/github-qna-webapp-react/.env:1]: pls delete this file

.env is in .gitignore, could you check why it's showing up?

Comment (adrianwyatt) [reply samples/apps/github-qna-webapp-react/.env:1]: done.

File: samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+                    SystemPromptDefaults.SystemIntentContinuationPrompt
+                })
+            );
+
+        var intentExtractionVariables = new ContextVariables();
+        intentExtractionVariables.Set("tokenLimit", historyTokenBudget.ToString(new NumberFormatInfo()));
+        intentExtractionVariables.Set("knowledgeCutoff", SystemPromptDefaults.KnowledgeCutoffDate);
+        intentExtractionVariables.Set("audience", context["audience"]);
+
+        var completionFunction = this._kernel.CreateSemanticFunction(
```
Comment (lemillermicrosoft) [inline samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs:55]: nit: Any reason to not initialize this once in the constructor?

File: samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+        var userIntent = await this.ExtractUserIntentAsync(context);
+        remainingToken -= this.EstimateTokenCount(userIntent);
+
+        context.Variables.Set("tokenLimit", remainingToken.ToString(new NumberFormatInfo()));
+        context.Variables.Set("contextTokenLimit", contextTokenLimit.ToString(new NumberFormatInfo()));
+        context.Variables.Set("knowledgeCutoff", SystemPromptDefaults.KnowledgeCutoffDate);
+        context.Variables.Set("userIntent", userIntent);
+        context.Variables.Set("audience", context["audience"]);
+
+        var completionFunction = this._kernel.CreateSemanticFunction(
```
Comment (lemillermicrosoft) [inline samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs:202]: Same as above, could these be created just once?

File: samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+        var remainingToken = Math.Min(
+            tokenLimit,
+            Math.Floor(contextTokenLimit * SystemPromptDefaults.MemoriesResponseContextWeight)
+        );
+
+        string memoryText = "";
+        var latestMessage = await this.GetLatestMemoryAsync(context);
+        if (latestMessage != null)
+        {
+            var results = context.Memory.SearchAsync("ChatMessages", latestMessage.Metadata.Text, limit: 1000);
```
Comment (lemillermicrosoft) [inline samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs:96]: nit: Would be good to have "ChatMessages" as a constant.

File: samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs (hunk 1 above)
Comment (adrianwyatt) [reply samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs:55]: Let's save this change for fit-and-finish. I'll create a tracking task.

File: samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs (hunk 2 above)
Comment (adrianwyatt) [reply samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs:202]: Let's save this change for fit-and-finish. I'll create a tracking task.

File: samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs (hunk 3 above)
Comment (adrianwyatt) [reply samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs:96]: Let's save this change for fit-and-finish. I'll create a tracking task.

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestApiOperation.cs [hunk 4]
```diff
@@ -37,7 +36,7 @@ internal class RestApiOperation
 
     /// <summary>
     /// The operation method - GET, POST, PUT, DELETE.
-    /// </summary>
+    /// </summary>o
```
Comment (lemillermicrosoft) [inline dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestApiOperation.cs:39]: Revert.

File: dotnet/src/SemanticKernel/Memory/IMemoryStore.cs [hunk 5]
```diff
@@ -114,7 +114,7 @@ public interface IMemoryStore
         CancellationToken cancel = default);
 
     /// <summary>
-    /// Gets the nearest match to the <see cref="Embedding"/> of type <see cref="float"/>.
+    /// Gets the nearest match to the <see cref="Embedding"/> of type <see cref="float"/>. Does not guaranteed that the collection exists.
```
[the remark this answers is not in the document]
Comment (adrianwyatt) [reply dotnet/src/SemanticKernel/Memory/IMemoryStore.cs:117]: Fixing grammer. It's easier to see with Intellisense when in the summary.

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+        await container.UpsertItemAsync(entity, cancellationToken: cancel, requestOptions: new ItemRequestOptions()
+        {
+            EnableContentResponseOnWrite = false,
+        });
+
+        return record.Key;
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<string> UpsertBatchAsync(string collectionName, IEnumerable<MemoryRecord> records, CancellationToken cancel = default)
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs:197]: Async-iterator 'CosmosDBMemoryStore.UpsertBatchAsync(string, IEnumerable<MemoryRecord>, CancellationToken)' has one or more parameters of type 'CancellationToken' but none of them is decorated with the 'EnumeratorCancellation' attribute, so the cancellation token parameter from the generated 'IAsyncEnumerable<>.GetAsyncEnumerator' will be unconsumed

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+            {
+                CosmosDBMemoryRecord record;
+
+                if (typeof(Stream).IsAssignableFrom(typeof(CosmosDBMemoryRecord)))
+                {
+                    record = ((CosmosDBMemoryRecord)(object)responseMessage.Content);
+                }
+                else
+                {
+                    record = await System.Text.Json.JsonSerializer.DeserializeAsync<CosmosDBMemoryRecord>(responseMessage.Content!, cancellationToken: cancel);
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs:141]: Converting null literal or possible null value to non-nullable type.

File: dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestApiOperation.cs (hunk 4 above)
Comment (adrianwyatt) [reply dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Model/RestApiOperation.cs:39]: will do

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs (hunk 7 above)
Comment (adrianwyatt) [reply dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs:141]: content non-null is checked in the using statement as well as when the response message succeeds. We can put some extra sauce around this in fit-and-finish. I'll add a tracking task.

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/Connectors.Memory.CosmosDB.csproj [hunk 8]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <RepoRoot>$([System.IO.Path]::GetDirectoryName($([MSBuild]::GetPathOfFileAbove('.gitignore', '$(MSBuildThisFileDirectory)'))))</RepoRoot>
+  </PropertyGroup>
+  <Import Project="$(RepoRoot)/dotnet/nuget/nuget-package.props" />
+  
+  <PropertyGroup>
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.Memory.CosmosDB/Connectors.Memory.CosmosDB.csproj:7]: I expect more details here like AssemblyName, RootNamespace and nuget details (PackageId, Title).

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryRecord.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.SemanticKernel.Memory;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace Connectors.Memory.CosmosDB;
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryRecord.cs:9]: Any reason not `Microsoft.SemanticKernel.Connectors.Memory.CosmosDB` like the others?

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+        {
+            this._log.LogInformation("Created database {0}", this._databaseName);
+        }
+        else if (response.StatusCode == HttpStatusCode.OK)
+        {
+            this._log.LogInformation("Database {0}", this._databaseName);
+        }
+        else
+        {
+            throw new CosmosException("Database does not exist and was not created", response.StatusCode, 0, this._databaseName, 0);
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs:58]: Is this part of `Microsoft.Azure.Cosmos`?

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+/// The embedding data persists between subsequent instances and has similarity search capability, handled by the client as Azure Cosmos DB is not a vector-native DB.
+/// </remarks>
+public class CosmosDBMemoryStore : IMemoryStore
+{
+    private Database _database;
+    private string _databaseName;
+    private ILogger _log;
+
+    /// <summary>
+    /// Create an instance of <see cref="CosmosDBMemoryStore"/>
```
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs:34]: nit: xmldoc details

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/Connectors.Memory.CosmosDB.csproj (hunk 8 above)
Comment (adrianwyatt) [reply dotnet/src/Connectors/Connectors.Memory.CosmosDB/Connectors.Memory.CosmosDB.csproj:7]: Looks like these got lost in the MemoryStore refactor. I'll re-add them back and disable nuget publishing for this project for now.

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryRecord.cs (hunk 9 above)
Comment (adrianwyatt) [reply dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryRecord.cs:9]: updating.

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs (hunk 10 above)
Comment (adrianwyatt) [reply dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs:58]: Yes.

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs (hunk 11 above)
Comment (adrianwyatt) [reply dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs:34]: will update

File: dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs (hunk 7 above)
Comment (lemillermicrosoft) [reply dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs:141]: I'd double request this one as it's a build warning that will impact everyone.

Comment (adrianwyatt) [reply dotnet/src/Connectors/Connectors.Memory.CosmosDB/CosmosDBMemoryStore.cs:141]: will update -standby

**Review by lemillermicrosoft — APPROVED** [review]
LGTM. Advise squash merge unless you are going to rebase the feature branch to remove the 37 merge commits.

**Pull request merged**